x-pack/apm-server/sampling/eventstorage/storage.go

```diff
@@ -0,0 +1,228 @@
+//
+// This must be called when the writer is no longer needed, in order to reclaim
+// resources.
+func (rw *ReadWriter) Close() {
+	rw.txn.Discard()
+}
+
+// Flush waits for preceding writes to be committed to storage.
+//
+// Flush must be called to ensure writes are committed to storage.
+// If Flush is not called before the writer is closed, then writes
+// may be lost.
+func (rw *ReadWriter) Flush() error {
+	err := rw.txn.Commit()
+	rw.txn = rw.s.db.NewTransaction(true)
+	rw.pendingWrites = 0
+	return err
+}
+
+// WriteTraceSampled records the tail-sampling decision for the given trace ID.
+func (rw *ReadWriter) WriteTraceSampled(traceID string, sampled bool) error {
+	key := []byte(traceID)
+	var meta uint8 = entryMetaTraceUnsampled
+	if sampled {
+		meta = entryMetaTraceSampled
+	}
+	entry := badger.NewEntry(key[:], nil).WithMeta(meta)
+	return rw.writeEntry(entry.WithTTL(rw.s.ttl))
+}
+
+// IsTraceSampled reports whether traceID belongs to a trace that is sampled
+// or unsampled. If no sampling decision has been recorded, IsTraceSampled
+// returns ErrNotFound.
+func (rw *ReadWriter) IsTraceSampled(traceID string) (bool, error) {
+	rw.readKeyBuf = append(rw.readKeyBuf[:0], traceID...)
+	item, err := rw.txn.Get(rw.readKeyBuf)
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return false, ErrNotFound
+		}
+		return false, err
+	}
+	return item.UserMeta() == entryMetaTraceSampled, nil
+}
+
+// WriteTransaction writes tx to storage.
+//
+// WriteTransaction may return before the write is committed to storage.
+// Call Flush to ensure the write is committed.
+func (rw *ReadWriter) WriteTransaction(tx *model.Transaction) error {
+	if tx.Sampled != nil && !*tx.Sampled {
+		return errors.New("transaction is not sampled")
+	}
+	key := append(append([]byte(tx.TraceID), ':'), tx.ID...)
+	data, err := rw.s.codec.EncodeTransaction(tx)
+	if err != nil {
+		return err
+	}
+	return rw.writeEvent(key[:], data, entryMetaTransaction)
+}
+
+// WriteSpan writes span to storage.
+//
+// WriteSpan may return before the write is committed to storage.
+// Call Flush to ensure the write is committed.
+func (rw *ReadWriter) WriteSpan(span *model.Span) error {
+	key := append(append([]byte(span.TraceID), ':'), span.ID...)
+	data, err := rw.s.codec.EncodeSpan(span)
+	if err != nil {
+		return err
+	}
+	return rw.writeEvent(key[:], data, entryMetaSpan)
+}
+
+func (rw *ReadWriter) writeEvent(key, value []byte, meta byte) error {
+	return rw.writeEntry(badger.NewEntry(key, value).WithMeta(meta).WithTTL(rw.s.ttl))
+}
+
+func (rw *ReadWriter) writeEntry(e *badger.Entry) error {
+	rw.pendingWrites++
+	err := rw.txn.SetEntry(e)
+	if err != badger.ErrTxnTooBig {
+		return err
+	}
+	if err := rw.Flush(); err != nil {
+		return err
+	}
+	return rw.txn.SetEntry(e)
+}
+
+// ReadEvents reads events with the given trace ID from storage into a batch.
+func (rw *ReadWriter) ReadEvents(traceID string, out *model.Batch) error {
```

Comment:
The ReadWriter type's doc comment already states the whole type is not safe for concurrent usage. Is there still something unclear?

Reply:
I would not necessarily expect a write action on a `ReadEvents` method, so when using and locking it in the caller, there might only be a read lock used and not a write lock. But one would probably look at the code and realize that, so it's ok.

Reply:
Okay, understood - I'll add a comment to ReadEvents stating that it may also cause writes. I don't think we'll end up with read/write locks as the event storage is write-heavy, and ReadEvents is only called periodically.